packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -0,0 +1,210 @@
+  final AWSApiConfig _config;
+
+  // Manages all incoming messages from server. Primarily handles messages related
+  // to the entire connection. E.g. connection_ack, connection_error, ka, error.
+  // Other events (for single subscriptions) rebroadcast to _rebroadcastController.
+  late final WebSocketChannel _channel;
+  late final StreamSubscription<WebSocketMessage> _subscription;
+  late final RestartableTimer _timeoutTimer;
+
+  // Re-broadcasts incoming messages for child streams (single GraphQL subscriptions).
+  // start_ack, data, error
+  final StreamController<WebSocketMessage> _rebroadcastController =
+      StreamController<WebSocketMessage>.broadcast();
+  Stream<WebSocketMessage> get _messageStream => _rebroadcastController.stream;
+
+  // TODO: Add connection error variable to throw in `init`.
+
+  // Futures to manage initial connection state.
+  Future<void>? _initFuture;
+  final Completer<void> _connectionReady = Completer<void>();
+
+  /// Fires when the connection is ready to be listened to, i.e.
+  /// after the first `connection_ack` message.
+  Future<void> get ready => _connectionReady.future;
+
+  /// {@macro amplify_api.web_socket_connection}
+  WebSocketConnection(this._config, this._authProviderRepo);
+
+  /// Connects _subscription stream to _onData handler.
+  @visibleForTesting
+  StreamSubscription<WebSocketMessage> getStreamSubscription(
+      Stream<dynamic> stream) {
+    return stream
+        .transform(const WebSocketMessageStreamTransformer())
+        .listen(_onData);
+  }
+
+  /// Connects WebSocket channel to _subscription stream but does not send connection
+  /// init message.
+  @visibleForTesting
+  Future<void> connect(Uri connectionUri) async {
+    _channel = WebSocketChannel.connect(
+      connectionUri,
+      protocols: webSocketProtocols,
+    );
+    _subscription = getStreamSubscription(_channel.stream);
+  }
+
+  /// Closes the WebSocket connection.
+  @override
+  void close() {
+    _subscription.cancel();
+    _channel.sink.close();
+  }
+
+  /// Initializes the connection.
+  ///
+  /// Connects to WebSocket, sends connection message and resolves future once
+  /// connection_ack message received from server.
+  Future<void> init() {
+    return _initFuture ??= _init();
+  }
+
+  Future<void> _init() async {
+    final connectionUri =
+        await generateConnectionUri(_config, _authProviderRepo);
+    await connect(connectionUri);
+
+    if (_connectionReady.isCompleted) return;
+    send(WebSocketConnectionInitMessage());
+    return ready;
+  }
+
+  /// Subscribes to the given GraphQL request. Returns the subscription object,
+  /// or throws an [Exception] if there's an error.
+  Stream<GraphQLResponse<T>> subscribe<T>(
+    GraphQLRequest<T> request,
+    void Function()? onEstablished,
+  ) {
+    if (!_connectionReady.isCompleted) {
+      init();
+    }
+
+    final subscriptionId = uuid();
+    return _messageStream
```

Comment:
Request: Can you verify the onDone() function triggers on the client side? When I was playing around with the code it didn't seem to trigger when I thought it might. Could be my assumptions were wrong, but let's make sure it's wired up correctly. 

Tests around these callback functions might be appropriate too.